crates/ra_ide/src/hover.rs

```diff
@@ -335,6 +346,277 @@ fn hover_for_definition(db: &RootDatabase, def: Definition) -> Option<Markup> {
+        Namespace::Types => (hashset!{"type", "struct", "enum", "mod", "trait", "union", "module"}, hashset!{}),
+        Namespace::Values => (hashset!{"value", "function", "fn", "method", "const", "static", "mod", "module"}, hashset!{"()"}),
+        Namespace::Macros => (hashset!{"macro"}, hashset!{"!"})
+    };
+);
+
+impl Namespace {
+    /// Extract the specified namespace from an intra-doc-link if one exists.
+    fn from_intra_spec(s: &str) -> Option<Self> {
+        NS_MAP
+            .iter()
+            .filter(|(_ns, (prefixes, suffixes))| {
+                prefixes
+                    .iter()
+                    .map(|prefix| {
+                        s.starts_with(prefix)
+                            && s.chars()
+                                .nth(prefix.len() + 1)
+                                .map(|c| c == '@' || c == ' ')
+                                .unwrap_or(false)
+                    })
+                    .any(|cond| cond)
+                    || suffixes
+                        .iter()
+                        .map(|suffix| {
+                            s.starts_with(suffix)
+                                && s.chars()
+                                    .nth(suffix.len() + 1)
+                                    .map(|c| c == '@' || c == ' ')
+                                    .unwrap_or(false)
+                        })
+                        .any(|cond| cond)
+            })
+            .map(|(ns, (_, _))| *ns)
+            .next()
+    }
+}
+
+// Strip prefixes, suffixes, and inline code marks from the given string.
+fn strip_prefixes_suffixes(mut s: &str) -> &str {
+    s = s.trim_matches('`');
+    NS_MAP.iter().for_each(|(_, (prefixes, suffixes))| {
+        prefixes.iter().for_each(|prefix| s = s.trim_start_matches(prefix));
+        suffixes.iter().for_each(|suffix| s = s.trim_end_matches(suffix));
+    });
+    s.trim_start_matches("@").trim()
+}
+
+/// Try to resolve path to local documentation via intra-doc-links (i.e. `super::gateway::Shard`).
+///
+/// See [RFC1946](https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md).
+fn try_resolve_intra(
+    db: &RootDatabase,
+    definition: &Definition,
+    link_text: &str,
+    link_target: &str,
+) -> Option<(String, String)> {
+    // Set link_target for implied shortlinks
+    let link_target =
+        if link_target.is_empty() { link_text.trim_matches('`') } else { link_target };
+
+    // Namespace disambiguation
+    let namespace = Namespace::from_intra_spec(link_target);
+
+    // Strip prefixes/suffixes
+    let link_target = strip_prefixes_suffixes(link_target);
+
+    // Parse link as a module path
+    let path = Path::parse(link_target).ok()?;
+    let modpath = ModPath::from_src(path, &Hygiene::new_unhygienic()).unwrap();
+
+    // Resolve it relative to symbol's location (according to the RFC this should consider small scopes
+    let resolver = definition.resolver(db)?;
```

Comment:
Rather than exposing resolver, we should have `fn resolve_doc_path(path: &str)` method on appropriate types in `code_model`.

Reply:
Do you think that's the right signature? It'd be inclined to use something like: `fn resolve_doc_path(path: &ModPath) -> PerNs` or `fn resolve_doc_path(path: &ModPath, ns: Namespace) -> ItemInNs`.

Reply:
It is the right signature -- the path in the doc comment is a string, it's hir's job to parse that string into a semantic repr (`ModPath`). Generally, everything above *hir* should be pretty dumb and don't try to interpret "components" of things (not that we are super successfull with this layering...)